extension/src/repository/model/collect.ts

```diff
@@ -252,32 +179,180 @@ export const collectSelected = (
   return acc
 }
 
-const isUncollectedChild = (
-  deleted: Set<string>,
-  deletedPath: string,
-  trackedPath: string
-): boolean => {
-  return !deleted.has(trackedPath) && isSameOrChild(deletedPath, trackedPath)
+const getAbsPath = (dvcRoot: string, relPath: string): string =>
+  resolve(dvcRoot, relPath)
+
+export const UndecoratedDataStatus = {
+  TRACKED_DECORATIONS: 'trackedDecorations',
+  UNCHANGED: 'unchanged',
+  UNTRACKED: 'untracked'
+} as const
+
+const AvailableDataStatus = Object.assign(
+  {} as const,
+  DecorationDataStatus,
+  UndecoratedDataStatus
+)
+
+export type Status =
+  typeof AvailableDataStatus[keyof typeof AvailableDataStatus]
+
+type DataStatusMapping = { [path: string]: Status }
+
+export type DataStatus = Record<Status, Set<string>>
+
+const getStatus = (
+  path: string,
+  original: DataStatusMapping,
+  withMissingAncestors: DataStatusMapping
+) => original[path] || withMissingAncestors[path]
+
+const addMissingWithAncestorStatus = (
+  withMissingAncestors: DataStatusMapping,
+  missingAncestors: Set<string>,
+  status: Status
+): void => {
+  for (const ancestor of missingAncestors) {
+    withMissingAncestors[ancestor] = status
+  }
 }
 
-const collectIfDeletedChild = (
-  deleted: Set<string>,
-  deletedPath: string,
-  trackedPath: string
+const addMissingAncestors = (
+  pathArray: string[],
+  original: DataStatusMapping,
+  withMissingAncestors: DataStatusMapping
 ): void => {
-  if (isUncollectedChild(deleted, deletedPath, trackedPath)) {
-    deleted.add(trackedPath)
+  const missingAncestors = new Set<string>()
+
+  for (let reverseIdx = pathArray.length; reverseIdx > 0; reverseIdx--) {
+    const path = getPath(pathArray, reverseIdx)
+
+    const status = getStatus(path, original, withMissingAncestors)
+    if (status) {
+      addMissingWithAncestorStatus(
+        withMissingAncestors,
+        missingAncestors,
+        status
+      )
+      return
+    }
+
+    missingAncestors.add(path)
   }
 }
 
-export const collectDeleted = (
-  deleted: Set<string>,
-  tracked: Set<string>
-): Set<string> => {
-  for (const trackedPath of tracked) {
-    for (const deletedPath of deleted) {
-      collectIfDeletedChild(deleted, deletedPath, trackedPath)
+const collectMissingAncestors = (originalMapping: {
+  [path: string]: Status
+}): DataStatusMapping => {
+  const withMissingAncestors: DataStatusMapping = {}
+
+  for (const [path, status] of Object.entries(originalMapping)) {
+    withMissingAncestors[path] = status
+    const pathArray = getPathArray(path)
+    addMissingAncestors(
+      pathArray.slice(0, -1),
+      originalMapping,
+      withMissingAncestors
+    )
+  }
+
+  return withMissingAncestors
+}
+
+const addToTracked = (
+  tracked: Set<string>,
+  absPath: string,
+  status: Status
+) => {
+  if (status === AvailableDataStatus.UNTRACKED) {
```

Comment:
[F] All of the statuses in `data status` are tracked with the exception of the `untracked` key.